lib/instrumentation.js

```diff
@@ -20,9 +28,51 @@ module.exports = function initialize(shim, Koa) {
       return original.apply(this, arguments)
     }
   })
+  shim.wrapReturn(Koa.prototype, 'createContext', wrapCreateContext)
+  function wrapCreateContext(shim, fn, fnName, context) {
+    Object.defineProperty(context, 'body', {
+      get: function getBody() {
+        return this.__NR_body
+      },
+      set: function setBody(val) {
+        shim.savePossibleTransactionName(this.req)
+        this.__NR_body = val
+      }
+    })
+  }
 }
 
 function wrapMiddleware(shim, middleware) {
+  var router = middleware.router
+  if (usingKoaRouter && router && router.stack && router.stack.length) {
+    var stack = router.stack
+    for (var i = 0; i < stack.length; ++i) {
+      var layer = stack[i]
+      var spec = {
+        route: layer.path,
+        type: shim.MIDDLEWARE,
+        next: shim.LAST,
```

Comment:
they are promise based, but the sequencing is bound up in `koa-compose` which has similar semantics to express' middleware sequencing.  the alternative will cause all middleware to be nested under siblings like the router segments are now.

i suppose it could be more accurate to have them nested like that and tie the end of timing to the resolution of the middleware promise, since you could have more work done in the middleware after the next call has resolved.

Reply:
wrapping the middleware based on their promises resolving will nest the names of the middleware matched incorrectly, so if i have sibling middleware mounted on `/:first` and `/:second` a response from `/:second` will mark the transaction as `/:first/:second` due to the middleware being nested in the trace.

Reply:
If you are not basing the middleware segment's duration off the promise resolving then middleware such as the one below are going to be recorded wrong. Is this true?

```js
router.use('/foo', async (ctx, next) => {
  await thisCreatesASegment()

  await next()

  await thisAlsoCreatesASegment()
})
```

Being able to do things in the middleware after calling next is a core design principal of Koa, so we must support this.

Reply:
a test case has been added for this scenario, and changes in the agent have been made to support this.